src/sparseml/pytorch/utils/module.py

```diff
@@ -987,13 +989,17 @@ def _runner_batch(
         )
 
         # optimizer / gradients update
-        if self._accumulated == self._num_accumulated_batches:
+        if (
+            self._accumulated == self._num_accumulated_batches
+            or self._accumulated == counter_len
+        ):
             if self.device_context.use_mixed_precision:
                 self._scaler.step(self._optimizer)
                 self._scaler.update()
             else:
                 self._optimizer.step(closure=self._optim_closure)
             self._accumulated = 0
+            self._optimizer.zero_grad()
```

Comment:
this changes the order of step/zero_grad right? before zero_grad would happen before step, and now it seems like it happens after it?

Reply:
I'm not sure why it is currently happening before. For `_num_accumulated_batches > 1` this would lead to issues, since it would clear the gradients from the previous updates before they were used to update. Maybe the check was meant to be `if self._accumulated == (self._num_accumulated_batches + 1):`. I suppose there might be a potential use for the gradients between `_runner_batch()` calls. I'll update to clear gradients at the beginning